// @middleware.io/node-apm/lib/src/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerErrorHandler = exports.getTracer = exports.getMeter = exports.setAttribute = exports.errorRecord = exports.error = exports.debug = exports.warn = exports.info = exports.sdkShutdown = exports.track = void 0;
const otel = require("@opentelemetry/api");
const logger_1 = require("./logger");
const config_1 = require("./config");
const profiler_1 = require("./profiler");
const config_2 = require("./config");
const healthcheck_1 = require("./healthcheck");
const tracer_collector_1 = require("./tracer-collector");
const errorhandler_1 = __importDefault(require("./errorhandler"));
const track = (newConfig = {}) => {
    const config = (0, config_1.init)(newConfig);
    (0, profiler_1.init)(config).then((r) => { });
    // Perform Health check to MW Agent
    if (!config_2.configDefault["isServerless"]) {
        // Perform in async way as (synchronous import)
        (0, healthcheck_1.performHealthCheck)(config === null || config === void 0 ? void 0 : config.host).then((r) => { });
    }
};
exports.track = track;
const sdkShutdown = () => {
    return (0, tracer_collector_1.shutdown)();
};
exports.sdkShutdown = sdkShutdown;
const info = (message, attributes = {}) => {
    (0, logger_1.log)("INFO", message, attributes);
};
exports.info = info;
const warn = (message, attributes = {}) => {
    (0, logger_1.log)("WARN", message, attributes);
};
exports.warn = warn;
const debug = (message, attributes = {}) => {
    (0, logger_1.log)("DEBUG", message, attributes);
};
exports.debug = debug;
const error = (message, attributes = {}) => {
    (0, logger_1.log)("ERROR", message, attributes);
};
exports.error = error;
const errorRecord = (e) => {
    if (otel === null || otel === void 0 ? void 0 : otel.context) {
        const span = otel.trace.getSpan(otel.context.active());
        if (span) {
            span.recordException(e);
            span.setStatus({ code: otel.SpanStatusCode.ERROR, message: String(e) });
        }
    }
};
exports.errorRecord = errorRecord;
const setAttribute = (name, value) => {
    if (otel === null || otel === void 0 ? void 0 : otel.context) {
        const span = otel.trace.getSpan(otel.context.active());
        if (span) {
            span.setAttribute(name, value);
        }
    }
};
exports.setAttribute = setAttribute;
const getMeter = (name, version, options) => {
    if (config_2.configDefault.meterProvider) {
        return config_2.configDefault.meterProvider.getMeter(config_2.configDefault.serviceName);
    }
    return otel.meter.getMeter(config_2.configDefault.serviceName);
};
exports.getMeter = getMeter;
const getTracer = (name, version, options) => {
    return otel.trace.getTracer(config_2.configDefault.serviceName);
};
exports.getTracer = getTracer;
// Function to register the error handler
function registerErrorHandler(app) {
    app.use(errorhandler_1.default);
}
exports.registerErrorHandler = registerErrorHandler;
